File: csharp/Euler/p0020.cs
/*
Project Euler Problem 20

Problem:

n! means n × (n − 1) × ... × 3 × 2 × 1

For example, 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800,
and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.

Find the sum of the digits in the number 100!
*/
using System;

namespace Euler
{
    public class p0020 : IEuler
    {
        public object Answer()
        {
            ulong[] numbers = new ulong[10];
            const ulong ten17 = 100000000000000000;
            numbers[0] = 1;
            for (byte i = 2; i <= 100; i++)
            {
                for (byte j = 0; j < 10; j++)
                {
                    numbers[j] *= i;
                }
                for (byte j = 0; j < 9; j++)
                {
                    if (numbers[j] > ten17)
                    {
                        numbers[j + 1] += numbers[j] / ten17;
                        numbers[j] %= ten17;
                    }
                }
            }
            ulong answer = 0;
            ulong power = 1;
            for (byte i = 0; i < 19; i++)
            {
                for (byte j = 0; j < 10; j++)
                {
                    ulong value = numbers[j] / power;
                    answer += value % 10;
                }
                power *= 10;
            }
            return (ushort)answer;
        }
    }
}

File: java/src/main/java/euler/p0020.java
/*
Project Euler Problem 34

This ended up being a filtering problem. The problem with my solution is that I
am not satisfied with my filter at all. I feel like there is a more efficient
way to go about it.

Problem:

145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.

Find the sum of all numbers which are equal to the sum of the factorial of
their digits.

Note: as 1! = 1 and 2! = 2 are not sums they are not included.
*/
package euler;

public class p0020 implements IEuler {
    @Override
    public Object answer() {
        long[] numbers = new long[11];
        long ten16 = 10000000000000000L;
        numbers[0] = 1;
        for (byte i = 2; i <= 100; i++) {
            for (byte j = 0; j < 10; j++)
                numbers[j] *= i;
            for (byte j = 0; j < 9; j++) {
                if (numbers[j] > ten16) {
                    numbers[j + 1] += numbers[j] / ten16;
                    numbers[j] %= ten16;
                }
            }
        }
        long answer = 0;
        long power = 1;
        for (byte i = 0; i < 19; i++) {
            for (byte j = 0; j < 10; j++)
                answer += (numbers[j] / power) % 10;
            power *= 10;
        }
        return (short) answer;
    }
}
